### ISSUE TITLE
Client NPE in tablist

### ISSUE THREAD

**Issue by CCCCarsten** [CitizensDev/Citizens2#1360]
Hi,
server: Bungeecord (latest), Spigot 1.12.2 with ProtocolSupport (latest, 1.8.8 - 1.12.2).
Client: 1.12.2 version
It seems that citizens causes an issue on the client side:

Error executing task
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at h.a(SourceFile:47)
	at bib.az(SourceFile:991)
	at bib.a(SourceFile:419)
	at net.minecraft.client.main.Main.main(SourceFile:123)
Caused by: java.lang.NullPointerException
	at brz.a(SourceFile:482)
	at ic.a(SourceFile:67)
	at ic.a(SourceFile:14)
	at hv$1.run(SourceFile:13)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at h.a(SourceFile:46)
	... 3 more

This error appears near every second and causes lags on the client side.
Logfile of server dooesn't show errors.

**Comment by seema84** [comment]
can confirm

**Comment by seema84** [comment]
[main/FATAL]: Error executing task
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:1.8.0_25]
	at java.util.concurrent.FutureTask.get(FutureTask.java:192) ~[?:1.8.0_25]
	at net.minecraft.util.Util.func_181617_a(SourceFile:47) [h.class:?]
	at net.minecraft.client.Minecraft.func_71411_J(Minecraft.java:1086) [bib.class:?]
	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:397) [bib.class:?]
	at net.minecraft.client.main.Main.main(SourceFile:123) [Main.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_25]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_25]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_25]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[?:1.8.0_25]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
Caused by: java.lang.NullPointerException
	at net.minecraft.client.network.NetHandlerPlayClient.func_147237_a(NetHandlerPlayClient.java:558) ~[brz.class:?]
	at net.minecraft.network.play.server.SPacketSpawnPlayer.func_148833_a(SourceFile:67) ~[ic.class:?]
	at net.minecraft.network.play.server.SPacketSpawnPlayer.func_148833_a(SourceFile:14) ~[ic.class:?]
	at net.minecraft.network.PacketThreadUtil$1.run(PacketThreadUtil.java:20) ~[hv$1.class:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_25]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_25]
	at net.minecraft.util.Util.func_181617_a(SourceFile:46) ~[h.class:?]
	... 9 more

**Comment by seema84** [comment]
its only if citizens player npcs are loaded

**Comment by mcmonkey4eva** [comment]
Sounds like an issue with Bungee/ProtocolSupport?

We aren't equipped to diagnose or handle clientside errors, especially if they only replicate when sent through relay...

**Comment by seema84** [comment]
I don't use Bungee and ProtocolLib but this error stay. If i remove Citizens the error will stop.

**Comment by CCCCarsten** [comment]
Agree. It is very easy to close a ticket and more difficult to solve it.

**Comment by seema84** [comment]
I get this error also on a fresh Spigot Server (last Build) and only one Plugin (Citizens).
Thats definitive a Citizens problem.

**Comment by mcmonkey4eva** [comment]
Will reopen since it's replicated without interference.

Do note that it may be difficult to get any solution here, as we don't have any tooling to deal with clientside errors.

**Comment by seema84** [comment]
I pretty shure you can fix this. 
On some Server without Citizens but with NPC's i dont get this null errors.

**Comment by mcmonkey4eva** [comment]
Are you running Forge / similar modded client?

Can you get an error log generated with a Vanilla client?

**Comment by fullwall** [comment]
Seems likely to be this line `this.getPlayerInfo(packetIn.getUniqueId()).getGameProfile()`

**Comment by fullwall** [comment]
Can you paste your config

**Comment by seema84** [comment]
without forge:

> 

19:03:32 | bjb | Client thread | info | [CHAT] You created Hans.
-- | -- | -- | -- | --
19:03:33 | bib | Client thread | fatal | Error executing task java.util.concurrent.ExecutionException: java.lang.NullPointerException 	at java.util.concurrent.FutureTask.report(FutureTask.java:122) 	at java.util.concurrent.FutureTask.get(FutureTask.java:192) 	at h.a(SourceFile:47) 	at bib.az(SourceFile:991) 	at bib.a(SourceFile:419) 	at net.minecraft.client.main.Main.main(SourceFile:123) Caused by: java.lang.NullPointerException 	at brz.a(SourceFile:482) 	at ic.a(SourceFile:67) 	at ic.a(SourceFile:14) 	at hv$1.run(SourceFile:13) 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) 	at h.a(SourceFile:46) 	... 3 more

**Comment by seema84** [comment]
i get this error when a npc load/unload or i create a new one

@fullwall
its also on a testserver with a fresh config

**Comment by fullwall** [comment]
What about your saves.yml?

**Comment by seema84** [comment]
you cant reproduce and test it?

**Comment by fullwall** [comment]
No

**Comment by seema84** [comment]
https://pastebin.com/80Fja5NW

**Comment by seema84** [comment]
what do you need to reproduce this error?

**Comment by daipenger** [comment]
I have not looked at your source code. However I do have a strong suspicion what is going on and how to fix it (pretty much) after examining the client side of things:
You are sending a SPacketSpawnPlayer, immediately followed by SPacketPlayerListItem with a REMOVE_PLAYER action, am I correct? (that's how these packets have been renamed in forge, but it should be somewhat equivalent in your code)
It turns out that the client code apparently buffers all packets and executes them with multiple threads/not in order. (didn't exactly check that, just an educated guess from the stacktrace) So in some cases the SPacketPlayerListItem is processed before SPacketSpawnPlayer which removes network information that the processing method of the SPacketSpawnPlayer needs. Therefore the NPE.

So, as a workaround can you try out delaying the SPacketPlayerListItem by one tick?

**Comment by mibby** [comment]
I can confirm I am also experiencing this issue with Citizens. Error is coming from the client logs and repeatedly occurs hundreds of times while online.

```
[20:29:11] [main/FATAL] [net.minecraft.client.Minecraft]: Error executing task
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:1.8.0_25]
	at java.util.concurrent.FutureTask.get(FutureTask.java:192) ~[?:1.8.0_25]
	at net.minecraft.util.Util.func_181617_a(SourceFile:47) [h.class:?]
	at net.minecraft.client.Minecraft.func_71411_J(Minecraft.java:1086) [bib.class:?]
	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:397) [bib.class:?]
	at net.minecraft.client.main.Main.main(SourceFile:123) [Main.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_25]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_25]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_25]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[?:1.8.0_25]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
Caused by: java.lang.NullPointerException
	at net.minecraft.client.network.NetHandlerPlayClient.func_147237_a(NetHandlerPlayClient.java:558) ~[brz.class:?]
	at net.minecraft.network.play.server.SPacketSpawnPlayer.func_148833_a(SourceFile:67) ~[ic.class:?]
	at net.minecraft.network.play.server.SPacketSpawnPlayer.func_148833_a(SourceFile:14) ~[ic.class:?]
	at net.minecraft.network.PacketThreadUtil$1.run(PacketThreadUtil.java:20) ~[hv$1.class:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_25]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_25]
	at net.minecraft.util.Util.func_181617_a(SourceFile:46) ~[h.class:?]
	... 9 more
```

Client is on MC 1.12.2.
Paper dev 1441 (Spigot 1.12.2)
Citizens dev 1545.

**Comment by fullwall** [comment]
The latest build uses @daipenger's suggestion. Does this help the issue?

**Comment by seema84** [comment]
no don't help

**Comment by daipenger** [comment]
I have now done further and more sound analysis why that error occurs. Turns out that my first conclusion was incorrect. Recall that I said that packets are processed concurrently which is true as the network socket puts packets in a queue that are then processed by the main minecraft thread. But that does also mean that the processing of packets is still only one thread and therefore in order.
So https://github.com/CitizensDev/Citizens2/commit/ecd6df2e1548f1436b1a29c5c1cfc36e697d18ef is sadly totally useless.

I now completely monitored all SPacketPlayerListItem and SPacketSpawnPlayer packets. (Packet numbers 0x2E and 0x05, respectively) and found the culprit:

**Citizens does not obey the protocol specs!**

**Normally you expect sequences of the following form:**
**(1)** SPacketPlayerListItem with the add player action (action number 0)
**(2)** SPacketSpawnPlayer (which needs the saved state from (1)! If (1) wasn't sent, the NPE will happen! See http://wiki.vg/Protocol#Spawn_Player)
**(3)** SPacketPlayerListItem with the remove player action (action number 4)

Citizens sometimes (on apparently always the same NPCs, so there must be some NPC setting that causes it) does not send a (1), but instead a SPacketPlayerListItem with an update gamemode action. (action number 1)
In addition I also found out that Citizens sends (3) twice. If you can find out why, you should fix that, too.

I may continue to dig a bit further but I now hope that it is enough for you to fix that bug once and for all.

**Comment by mcmonkey4eva** [comment]
@daipenger this is not the fault of game mode packet -> that is only sent if you've set the gamemode explicitly by command.

I've gotten this error replicating locally, and have confirmed it still happens when game mode is not specifically set.

**Comment by mcmonkey4eva** [comment]
Replication notes:

Client error:
```
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at h.a(SourceFile:47)
	at bib.az(SourceFile:991)
	at bib.a(SourceFile:419)
	at net.minecraft.client.main.Main.main(SourceFile:123)
Caused by: java.lang.NullPointerException
	at brz.a(SourceFile:482)
	at ic.a(SourceFile:67)
	at ic.a(SourceFile:14)
	at hv$1.run(SourceFile:13)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at h.a(SourceFile:46)
	... 3 more
```

Timing aligns with send of two `PacketPlayOutPlayerInfo{action=REMOVE_PLAYER,uuid=<mynpc>}` packets.
Which were sent when player-type NPC was removed.

Note that only player spawn and player info packets were being tracked at that point. If any other packet was involved or sent in that time, i didn't see it.

**Comment by mcmonkey4eva** [comment]
Spent about 10 hours just now figuring this out.
Eventually traced it down to the Minecraft internals improperly sending the spawn packet way too early (during the initial spawn sequence), rather than waiting for the packet send that Citizens will do later.
(Which means it sent twice: once brokenly, then once validly).

I have created a fix in #1525 that patches this by blocking the initial missend.

I've confirmed the error no longer occurs with this patch.

**Comment by seema84** [comment]
"I have created a fix in #1525 that patches this by blocking the initial missend.
I've confirmed the error no longer occurs with this patch."

Thats not true. This error stay in all versions including up to #1​54​8

**Comment by seema84** [comment]
"I've confirmed the error no longer occurs with this patch."

I can't confirm. I use the build #1​54​8 and get this error when i create a new player npc or move in a world with existing npc's (when player are load or unload).

**Comment by mcmonkey4eva** [comment]
There is not a release build of my patch currently available. The fix is pending at  #1525
